File: ios/Classes/vungle/VungleInterstitialAdManager.swift
import Flutter
import UIKit
import VungleAdsSDK

class VungleInterstitialAdManager: NSObject {
    private let placementId: String
    private var eventSink: FlutterEventSink?
    private var interstitialAd: VungleInterstitial?
    
    init(placementId: String, eventSink: FlutterEventSink?) {
        self.placementId = placementId
        self.eventSink = eventSink
        super.init()
    }
    
    func loadAd() {
        print("[VungleInterstitialAd] Loading ad for placement: \(placementId)")
        
        interstitialAd = VungleInterstitial(placementId: placementId)
        interstitialAd?.delegate = self
        interstitialAd?.load()
    }
    
    func showAd() {
        print("[VungleInterstitialAd] Showing ad for placement: \(placementId)")
        
        guard let ad = interstitialAd, ad.canPlayAd() else {
            print("[VungleInterstitialAd] Ad cannot be played")
            sendEvent("onAdFailedToPlay", additionalData: [
                "error": ["code": -1, "message": "Ad cannot be played"]
            ])
            return
        }
        
        if let rootViewController = UIApplication.shared.keyWindow?.rootViewController {
            ad.present(with: rootViewController)
        }
    }
    
    func canPlayAd() -> Bool {
        return interstitialAd?.canPlayAd() ?? false
    }
    
    func dispose() {
        interstitialAd?.delegate = nil
        interstitialAd = nil
    }
    
    private func sendEvent(_ eventType: String, additionalData: [String: Any] = [:]) {
        var eventData: [String: Any] = [
            "event": eventType,
            "placementId": placementId
        ]
        eventData.merge(additionalData) { (_, new) in new }
        eventSink?(eventData)
    }
}

// MARK: - VungleInterstitialDelegate

extension VungleInterstitialAdManager: VungleInterstitialDelegate {
    func interstitialAdDidLoad(_ interstitial: VungleInterstitial) {
        print("[VungleInterstitialAd] onAdLoaded: \(placementId)")
        sendEvent("onAdLoaded")
    }
    
    func interstitialAdDidFailToLoad(_ interstitial: VungleInterstitial, withError: NSError) {
        print("[VungleInterstitialAd] onAdLoadFailed: \(placementId), error: \(withError.localizedDescription)")
        sendEvent("onAdLoadFailed", additionalData: [
            "error": ["code": withError.code, "message": withError.localizedDescription]
        ])
    }
    
    func interstitialAdWillPresent(_ interstitial: VungleInterstitial) {
        print("[VungleInterstitialAd] onAdWillPresent: \(placementId)")
    }
    
    func interstitialAdDidPresent(_ interstitial: VungleInterstitial) {
        print("[VungleInterstitialAd] onAdShowed: \(placementId)")
        sendEvent("onAdShowed")
    }
    
    func interstitialAdDidFailToPresent(_ interstitial: VungleInterstitial, withError: NSError) {
        print("[VungleInterstitialAd] onAdFailedToPlay: \(placementId), error: \(withError.localizedDescription)")
        sendEvent("onAdFailedToPlay", additionalData: [
            "error": ["code": withError.code, "message": withError.localizedDescription]
        ])
    }
    
    func interstitialAdDidTrackImpression(_ interstitial: VungleInterstitial) {
        print("[VungleInterstitialAd] onAdImpression: \(placementId)")
        sendEvent("onAdImpression")
    }
    
    func interstitialAdDidClick(_ interstitial: VungleInterstitial) {
        print("[VungleInterstitialAd] onAdClicked: \(placementId)")
        sendEvent("onAdClicked")
    }
    
    func interstitialAdWillLeaveApplication(_ interstitial: VungleInterstitial) {
        print("[VungleInterstitialAd] onAdWillLeaveApplication: \(placementId)")
    }
    
    func interstitialAdWillClose(_ interstitial: VungleInterstitial) {
        print("[VungleInterstitialAd] onAdWillClose: \(placementId)")
    }
    
    func interstitialAdDidClose(_ interstitial: VungleInterstitial) {
        print("[VungleInterstitialAd] onAdDismissed: \(placementId)")
        sendEvent("onAdDismissed")
    }
}

File: android/src/main/kotlin/com/example/multi_ads/vungle/VungleInterstitialAdManager.kt
package com.example.multi_ads.vungle

import android.app.Activity
import android.util.Log
import io.flutter.plugin.common.EventChannel
import com.vungle.ads.AdConfig
import com.vungle.ads.BaseAd
import com.vungle.ads.InterstitialAd
import com.vungle.ads.InterstitialAdListener
import com.vungle.ads.VungleError

class VungleInterstitialAdManager(
    private val activity: Activity,
    private val placementId: String,
    private val eventSink: EventChannel.EventSink?
) : InterstitialAdListener {

    companion object {
        const val TAG = "VungleInterstitialAd"
    }

    private var interstitialAd: InterstitialAd? = null

    fun loadAd() {
        Log.d(TAG, "Loading interstitial ad for placement: $placementId")
        
        interstitialAd = InterstitialAd(activity, placementId, AdConfig().apply {
            adOrientation = AdConfig.AUTO_ROTATE
        }).apply {
            adListener = this@VungleInterstitialAdManager
            load()
        }
    }

    fun showAd() {
        Log.d(TAG, "Showing interstitial ad for placement: $placementId")
        if (interstitialAd?.canPlayAd() == true) {
            interstitialAd?.play(activity)
        } else {
            Log.w(TAG, "Interstitial ad cannot be played")
            sendEvent("onAdFailedToPlay", mapOf(
                "error" to mapOf("code" to -1, "message" to "Ad cannot be played")
            ))
        }
    }

    fun canPlayAd(): Boolean {
        return interstitialAd?.canPlayAd() ?: false
    }

    fun dispose() {
        interstitialAd?.adListener = null
        interstitialAd = null
    }

    private fun sendEvent(eventType: String, additionalData: Map<String, Any?> = emptyMap()) {
        activity.runOnUiThread {
            val eventData = mutableMapOf<String, Any?>(
                "event" to eventType,
                "placementId" to placementId
            )
            eventData.putAll(additionalData)
            eventSink?.success(eventData)
        }
    }

    // InterstitialAdListener callbacks
    override fun onAdLoaded(baseAd: BaseAd) {
        Log.d(TAG, "onAdLoaded: $placementId")
        sendEvent("onAdLoaded")
    }

    override fun onAdStart(baseAd: BaseAd) {
        Log.d(TAG, "onAdStart: $placementId")
        sendEvent("onAdShowed")
    }

    override fun onAdImpression(baseAd: BaseAd) {
        Log.d(TAG, "onAdImpression: $placementId")
        sendEvent("onAdImpression")
    }

    override fun onAdEnd(baseAd: BaseAd) {
        Log.d(TAG, "onAdEnd: $placementId")
        sendEvent("onAdDismissed")
    }

    override fun onAdClicked(baseAd: BaseAd) {
        Log.d(TAG, "onAdClicked: $placementId")
        sendEvent("onAdClicked")
    }

    override fun onAdLeftApplication(baseAd: BaseAd) {
        Log.d(TAG, "onAdLeftApplication: $placementId")
    }

    override fun onAdFailedToLoad(baseAd: BaseAd, adError: VungleError) {
        Log.e(TAG, "onAdFailedToLoad: $placementId, error: ${adError.localizedMessage}")
        sendEvent("onAdLoadFailed", mapOf(
            "error" to mapOf("code" to adError.code, "message" to adError.localizedMessage)
        ))
    }

    override fun onAdFailedToPlay(baseAd: BaseAd, adError: VungleError) {
        Log.e(TAG, "onAdFailedToPlay: $placementId, error: ${adError.localizedMessage}")
        sendEvent("onAdFailedToPlay", mapOf(
            "error" to mapOf("code" to adError.code, "message" to adError.localizedMessage)
        ))
    }
}
